### PR TITLE
Add iai support and benchmarks for performance comparison between new PRs and main

### PR DESCRIPTION
**Pull request by xqft**
# Add iai support and benchmarks for performance comparison between new PRs and main

## Description

This PR refactors all benchmarks for adding support for iai so it can be used to detect regressions in the CI. Two new CI jobs were added for this purpose. Now on every PR iai will execute some benchmarks and compare them to the main branch. This PR was made taking [cairo-rs](https://github.com/lambdaclass/cairo-rs) as reference.

This also includes a small refactor of jobs, and including Apple sillicon benches given the recent merge of #266.

This will solve #249 .

**Note for reviewers**: the iai jobs will fail because the main branch tests are not named the same as in this PR. After merging this should solve itself.

## Type of change

- [x] Benchmarks

### CODE REVIEW

**Comment by xqft** [comment]
Note for reviewers: the iai jobs will fail because the main branch tests are not named the same as in this PR. After merging this should solve itself.

File: .github/workflows/criterion_benchs.yml [hunk 1]
```diff
@@ -1,10 +1,8 @@
-name: Benchmarks 
+name: CI
```
Comment (mpaulucci) [inline .github/workflows/criterion_benchs.yml:1]: why this change?

File: .github/workflows/iai_benchs_main.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+name: CI
```
Comment (mpaulucci) [inline .github/workflows/iai_benchs_main.yml:1]: why two separate files for benchmark jobs?

File: .github/workflows/criterion_benchs.yml (hunk 1 above)
Comment (xqft) [reply .github/workflows/criterion_benchs.yml:1]: To be consistent with the rest of the CI checks

File: .github/workflows/criterion_benchs.yml [hunk 3]
```diff
@@ -23,7 +22,11 @@ jobs:
           override: true
     - uses: actions/checkout@v3
     - name: Run benchmark
-      run: cargo bench --bench all_benchmarks -- --output-format bencher | tee output.txt
+      run: |
```
Comment (mpaulucci) [inline .github/workflows/criterion_benchs.yml:25]: this means that if a new bench is added, it has to be added here?

Comment (xqft) [reply .github/workflows/criterion_benchs.yml:25]: Yes, if it makes sense to run them on here. For example this job is missing the `criterion_metal` benchs that should execute on a different job running on our M2s, and `criterion_field` which benchs nanosecond ops between field elements, whose results get totally drowned with the CI's generated noise (check out [here](https://lambdaclass.github.io/lambdaworks/bench/)).

Comment (mpaulucci) [reply .github/workflows/criterion_benchs.yml:25]: makes sense, thx!

**Review by mpaulucci — APPROVED** [review]

File: gpu/benches/functions/metal.rs
Comment (Oppen) [inline gpu/benches/functions/metal.rs:1]: I'm not sure IAI is a good fit here. I'm assuming this runs mostly in the GPU, right? This means emulating main memory and CPU cache behavior will be a bad predictor of wall clock time.

Comment (xqft) [reply gpu/benches/functions/metal.rs:1]: You're right! this won't work. I'll leave only criterion benches. Thanks!

Comment (xqft) [reply gpu/benches/functions/metal.rs:1]: Oooh wait I think I never added them in the first place! Because there's no way to compile the `metal` module in the CI yet.

Comment (mpaulucci) [reply gpu/benches/functions/metal.rs:1]: If you think with main you should be able to get meta in the ci

Comment (xqft) [reply gpu/benches/functions/metal.rs:1]: Solved in latest commits

**Comment by xqft** [comment]
Closed in favour of #268, will cherry-pick some things from here

File: .github/workflows/iai_benchs_main.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    - name: Install valgrind
+      run: |
+        sudo apt update
+        sudo apt-get install -y valgrind
+    - name: Install stable toolchain
+      uses: dtolnay/rust-toolchain@stable
+      with:
+          toolchain: stable
+    - uses: actions/checkout@v3
+    - name: Initialize iai cache for ${{ github.event.pull_request.base.sha }}
```
Comment (Oppen) [inline .github/workflows/iai_benchs_main.yml:21]: This was a mistake on my part in the original pipeline. It worked pretty much by miracle. The correct line is:
```suggestion
    - name: Initialize iai cache for ${{ github.sha }}
```
That's because for push to main there's not actual pull request event, and the valid source of your commit hash is the one from the base context.

Comment (Oppen) [reply .github/workflows/iai_benchs_main.yml:21]: Oh, it worked because it's just the message lol

**Review by Oppen — APPROVED** [review]

**Review by pablodeymo — APPROVED** [review]
LGTM
